realm-jni/src/util.cpp

```diff
@@ -271,63 +250,73 @@ string concat_stringdata(const char *message, StringData strData)
-    // For efficiency, if the incoming UTF-8 string is sufficiently
-    // small, we will attempt to store the UTF-16 output into a stack
-    // allocated buffer of static size. Otherwise we will have to
-    // dynamically allocate the output buffer after calculating its
-    // size.
+jstring to_jstring(JNIEnv* env, StringData str) {
+    // Input is UTF-8 and output is UTF-16. Invalid UTF-8 input is silently
+    // converted to Unicode replacement characters.
+    // We use a small fixed size stack-allocated output buffer to avoid the cost
+    // of dynamic allocation for short input strings. If this buffer turns out
+    // to be too small, we proceed by calculating an estimate for the actually
+    // required output buffer size, and then allocate the buffer dynamically.
 
     const size_t stack_buf_size = 48;
     jchar stack_buf[stack_buf_size];
-    std::unique_ptr<jchar[]> dyn_buf;
-
-    const char* in_begin = str.data();
-    const char* in_end   = str.data() + str.size();
-    jchar* out_begin = stack_buf;
-    jchar* out_curr  = stack_buf;
-    jchar* out_end   = stack_buf + stack_buf_size;
-
-    typedef Utf8x16<jchar, JcharTraits> Xcode;
-
-    if (str.size() <= stack_buf_size) {
-        size_t retcode = Xcode::to_utf16(in_begin, in_end, out_curr, out_end);
-        if (retcode != 0)
-            throw runtime_error(string_to_hex("Failure when converting short string to UTF-16",  str, in_begin, in_end, out_curr, out_end, size_t(0), retcode));
-        if (in_begin == in_end)
-            goto transcode_complete;
-    }
-
-    {
-        const char* in_begin2 = in_begin;
-        size_t error_code;
-        size_t size = Xcode::find_utf16_buf_size(in_begin2, in_end, error_code);
-        if (in_begin2 != in_end) 
-            throw runtime_error(string_to_hex("Failure when computing UTF-16 size", str, in_begin, in_end, out_curr, out_end, size, error_code));
-        if (int_add_with_overflow_detect(size, stack_buf_size))
-            throw runtime_error("String size overflow");
-        dyn_buf.reset(new jchar[size]);
-        out_curr = copy(out_begin, out_curr, dyn_buf.get());
-        out_begin = dyn_buf.get();
-        out_end   = dyn_buf.get() + size;
-        size_t retcode = Xcode::to_utf16(in_begin, in_end, out_curr, out_end);
-        if (retcode != 0) 
-            throw runtime_error(string_to_hex("Failure when converting long string to UTF-16", str, in_begin, in_end, out_curr, out_end, size_t(0), retcode));
-        REALM_ASSERT(in_begin == in_end);
+    Buffer <jchar> dyn_buf;
+
+    const char *in = str.data();
+    const char *in_end = in + str.size();
+    jchar *out = stack_buf;
+    jchar *out_begin = out;
+    jchar *out_end = out_begin + stack_buf_size;
+
+    for (; ;) {
+        typedef Utf8x16 <jchar, JcharTraits> Xcode;
+        Xcode::to_utf16(in, in_end, out, out_end);
+        bool end_of_input = in == in_end;
+        if (end_of_input)
+            break;
+        bool bad_input = out != out_end;
```

Comment:
@kspangsege Thank you for confirming that.

You might have noticed that to_utf16() in utf8.hpp has a bug whose return value is different from what the comment says.
The comment says "true if all input transcoded or ...
                                 false if transcoding stopped due to invalidinput."
In fact, to_utf16() returns the opposite.
It seems that last line of to_utif16() shold be "return !invalid;"